Src/HALAL/Services/Encoder/Encoder.cpp

```diff
@@ -19,69 +22,107 @@ optional<uint8_t> Encoder::inscribe(Pin& pin1, Pin& pin2){
 
-	Encoder::id_manager.pop_front();
 	return id;
 }
 
+void Encoder::start(){
+	for_each(Encoder::registered_encoder.begin(), Encoder::registered_encoder.end(),
+			[](pair<uint8_t, pair<Pin, Pin>> iter) {
+				Encoder::init(Encoder::pin_timer_map[iter.second]);
+			}
+	);
+}
 
 void Encoder::turn_on(uint8_t id){
-	if (not Encoder::service_ids.contains(id)) {
+	if (not Encoder::registered_encoder.contains(id)) {
 		return; //TODO: error handler
 	}
 
-	TIM_HandleTypeDef* timer = pin_timer_map[service_ids[id]];
+	TIM_HandleTypeDef* timer = pin_timer_map[registered_encoder[id]].first;
 
 	if (HAL_TIM_Encoder_GetState(timer) == HAL_TIM_STATE_RESET) {
-		//TODO: Exception handle, Error (Encoder not initialized)
-		return;
+		return; //TODO: Exception handle, Error (Encoder not initialized)
+
 	}
 
 	if (HAL_TIM_Encoder_Start(timer, TIM_CHANNEL_ALL) != HAL_OK) {
-		//TODO: Exception handle, Warning (Error starting encoder)
+		return; //TODO: Exception handle, Warning (Error starting encoder)
 	}
 }
 
 void Encoder::turn_off(uint8_t id){
-	if (not Encoder::service_ids.contains(id)) {
+	if (not Encoder::registered_encoder.contains(id)) {
 		return; //TODO: error handler
 	}
 
-	TIM_HandleTypeDef* timer = pin_timer_map[service_ids[id]];
+	TIM_HandleTypeDef* timer = pin_timer_map[registered_encoder[id]].first;
 
 	if (HAL_TIM_Encoder_Stop(timer, TIM_CHANNEL_ALL) != HAL_OK) {
 		//TODO: Exception handle, Warning (Error stopping encoder)
 	}
 }
 
 void Encoder::reset(uint8_t id){
-	if (not Encoder::service_ids.contains(id)) {
+	if (not Encoder::registered_encoder.contains(id)) {
 		return; //TODO: error handler
 	}
 
-	TIM_HandleTypeDef* timer =  pin_timer_map[service_ids[id]];
+	TIM_HandleTypeDef* timer =  pin_timer_map[registered_encoder[id]].first;
 
-	timer->Instance->CNT = 0;
+	timer->Instance->CNT = UINT16_MAX / 2;
 }
 
 optional<uint32_t> Encoder::get_counter(uint8_t id){
-	if (not Encoder::service_ids.contains(id)) {
+	if (not Encoder::registered_encoder.contains(id)) {
 		return nullopt; //TODO: error handler
 	}
 
-	TIM_HandleTypeDef* timer = pin_timer_map[service_ids[id]];
+	TIM_HandleTypeDef* timer = pin_timer_map[registered_encoder[id]].first;
 
 	return timer->Instance->CNT;
 }
 
 optional<bool> Encoder::get_direction(uint8_t id){
-	if (not Encoder::service_ids.contains(id)) {
+	if (not Encoder::registered_encoder.contains(id)) {
 		return nullopt; //TODO: error handler
 	}
 
-	TIM_HandleTypeDef* timer =  pin_timer_map[service_ids[id]];
+	TIM_HandleTypeDef* timer =  pin_timer_map[registered_encoder[id]].first;
 
 	return ((timer->Instance->CR1 & 0b10000) >> 4);
 }
+
+void Encoder::init(pair<TIM_HandleTypeDef*, TIM_TypeDef*> encoder){
+	  TIM_Encoder_InitTypeDef sConfig = {0};
+	  TIM_MasterConfigTypeDef sMasterConfig = {0};
+
+	  encoder.first->Instance = encoder.second;
+	  encoder.first->Init.Prescaler = 0;
+	  encoder.first->Init.CounterMode = TIM_COUNTERMODE_UP;
+	  encoder.first->Init.Period = 65535;
```

Comment:
```suggestion
	  encoder.first->Instance = encoder.second;
	  encoder.first->Init.Prescaler = timer.prescaler;
	  encoder.first->Init.CounterMode = TIM_COUNTERMODE_UP;
	  encoder.first->Init.Period = timer.period;
```
Si utilizaras timer_peripheral puedes configurar el timer en el runes y luego aqui indicar la configuracion.

Reply:
Eso es constante.